datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+                    };
+                    let new_left =
+                        adjust_input_keys_down_recursively(left.clone(), left_keys)?;
+                    let new_right =
+                        adjust_input_keys_down_recursively(right.clone(), right_keys)?;
+                    Ok(Arc::new(HashJoinExec::try_new(
+                        new_left,
+                        new_right,
+                        new_join_on,
+                        filter.clone(),
+                        join_type,
+                        PartitionMode::Partitioned,
+                        null_equals_null,
+                    )?))
+                } else {
+                    let new_left = adjust_input_keys_down_recursively(
+                        left.clone(),
+                        join_key_pairs.left_keys,
+                    )?;
+                    let new_right = adjust_input_keys_down_recursively(
+                        right.clone(),
+                        join_key_pairs.right_keys,
+                    )?;
+                    Ok(Arc::new(HashJoinExec::try_new(
+                        new_left,
+                        new_right,
+                        on.clone(),
+                        filter.clone(),
+                        join_type,
+                        PartitionMode::Partitioned,
+                        null_equals_null,
+                    )?))
+                }
+            }
+            PartitionMode::CollectLeft => {
+                let new_left = adjust_input_keys_down_recursively(left.clone(), vec![])?;
+                let new_right = match join_type {
+                    JoinType::Inner | JoinType::Right => try_push_required_to_right(
+                        parent_required,
+                        right.clone(),
+                        left.schema().fields().len(),
+                    )?,
+                    JoinType::RightSemi | JoinType::RightAnti => {
+                        adjust_input_keys_down_recursively(
+                            right.clone(),
+                            parent_required.clone(),
+                        )?
+                    }
+                    JoinType::Left
+                    | JoinType::LeftSemi
+                    | JoinType::LeftAnti
+                    | JoinType::Full => {
+                        adjust_input_keys_down_recursively(right.clone(), vec![])?
+                    }
+                };
+
+                Ok(Arc::new(HashJoinExec::try_new(
+                    new_left,
+                    new_right,
+                    on.clone(),
+                    filter.clone(),
+                    join_type,
+                    PartitionMode::CollectLeft,
+                    null_equals_null,
+                )?))
+            }
+        }
+    } else if let Some(CrossJoinExec { left, right, .. }) =
+        plan_any.downcast_ref::<CrossJoinExec>()
+    {
+        let new_left = adjust_input_keys_down_recursively(left.clone(), vec![])?;
```

Comment:
Since the current implementation for the `CrossJoinExec` is similar to the `CollectLeft`, we don't adjust for the left side.